projects/rulesets-base/src/rule-runner/rule-runner.ts

```diff
@@ -189,4 +209,99 @@ export class RuleRunner {
 
     return [...externalResults, ...specificationResults, ...endpointResults];
   }
+
+  async runRules(inputs: {
+    context: any;
+    diffs: ObjectDiff[];
+    fromSpec: OpenAPIV3.Document;
+    toSpec: OpenAPIV3.Document;
+  }): Promise<RuleResult[]> {
+    const { context, fromSpec: beforeApiSpec, toSpec: afterApiSpec } = inputs;
+    const externalRules = this.rules.filter((rule) =>
+      ExternalRuleBase.isInstance(rule)
+    ) as ExternalRule[];
+    const rulesOrRulesets = this.rules.filter(
+      (rule) => !ExternalRuleBase.isInstance(rule)
+    ) as (Ruleset | Rule)[];
+
+    const externalResults: RuleResult[] = [];
+    for (const externalRule of externalRules) {
+      const results = await externalRule.runRulesV2(inputs);
+      externalResults.push(...results);
+    }
+
+    // TODO reimplement the rule runner so we don't need to generate legacy fact types here
+    const beforeFacts = traverseSpec(inputs.fromSpec);
+    const afterFacts = traverseSpec(inputs.toSpec);
+    const changelog = factsToChangelog(beforeFacts, afterFacts);
+
+    const openApiFactNodes = groupFacts({
+      beforeFacts,
+      afterFacts,
+      changes: changelog,
+    });
```

Comment:
Ideally, we'd be using the new traverser function here that I have and generate a fact tree, which would then be able to be used to determine which rules to run.

Since there's a lot of differences in this, I'd probably have to end up reimplementing the entire rule running, which if that was the case we should just refactor this.

If we want to:
- invest in performance
- invest in rule writing ergonomics

We shoud refactor this function